plugins/fpc/fu-fpc-device.c

```diff
@@ -0,0 +1,521 @@
+
+	if (!fu_fpc_device_dfu_cmd(self,
+				   FPC_CMD_DFU_GETSTATUS,
+				   0x0000,
+				   (guint8 *)dfu_status,
+				   sizeof(FuFpcDfuStatus),
+				   TRUE,
+				   FALSE,
+				   error)) {
+		g_prefix_error(error, "failed to get status: ");
+		return FALSE;
+	}
+
+	if ((dfu_status->bstatus != 0) || (dfu_status->bstate != FPC_DFU_IDLE)) {
+		/* device is not in correct status/state */
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "dfu status error [0x%x, 0x%x]",
+			    dfu_status->bstatus,
+			    dfu_status->bstate);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_update_init(FuFpcDevice *self, GError **error)
+{
+	FuFpcDfuStatus dfu_status = {0};
+
+	if (!fu_fpc_device_dfu_cmd(self,
+				   FPC_CMD_DFU_CLRSTATUS,
+				   0x0000,
+				   NULL,
+				   0,
+				   FALSE,
+				   FALSE,
+				   error)) {
+		g_prefix_error(error, "failed to clear status: ");
+		return FALSE;
+	}
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_fpc_device_check_dfu_status,
+				  FPC_DFU_MAX_ATTEMPTS,
+				  20,
+				  (gpointer)&dfu_status,
+				  error))
+		return FALSE;
+
+	if (dfu_status.bmax_payload_size)
+		self->max_block_size = FPC_FLASH_BLOCK_SIZE_4096;
+	else
+		self->max_block_size = FPC_FLASH_BLOCK_SIZE_DEFAULT;
+
+	return TRUE;
+}
+
+static void
+fu_fpc_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+	fu_string_append_kx(str, idt, "Max block size", self->max_block_size);
+	fu_string_append_kb(str, idt, "Is a moc device", self->is_moc_device);
+}
+
+static gboolean
+fu_fpc_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* sanity check */
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in runtime mode, skipping");
+		return TRUE;
+	}
+	if (!fu_fpc_device_dfu_cmd(self, FPC_CMD_DFU_DETACH, 0x0000, NULL, 0, FALSE, FALSE, error))
+		return FALSE;
+	fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* sanity check */
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in bootloader mode, skipping");
+		return TRUE;
+	}
+	if (!fu_fpc_device_fw_cmd(self, FPC_CMD_BOOT0, NULL, 0, FALSE, error))
+		return FALSE;
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_setup(FuDevice *device, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* setup */
+	if (!FU_DEVICE_CLASS(fu_fpc_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	self->is_moc_device = (fu_usb_device_get_pid(FU_USB_DEVICE(device)) >> 2 & 0x1) ? TRUE
+											: FALSE;
```

Comment:
is this **always** going to be true, or would a per-device quirk be a better idea?